pkg/storage/redis.go

```diff
@@ -0,0 +1,216 @@
+}
+
+func (b *RedisDB) Read(namespace, key string) ([]byte, error) {
+	nameSpaceKey := getRedisKey(namespace, key)
+	return b.db.Get(b.ctx, nameSpaceKey).Bytes()
+}
+
+func (b *RedisDB) ReadPrefix(namespace, prefix string) (map[string][]byte, error) {
+	namespacePrefix := getRedisKey(namespace, prefix)
+
+	keys, err := readAllKeys(namespacePrefix, b)
+	if err != nil {
+		return nil, errors.Wrap(err, "read all keys error")
+	}
+
+	return readAll(keys, b)
+}
+
+func (b *RedisDB) ReadAll(namespace string) (map[string][]byte, error) {
+	keys, err := readAllKeys(namespace, b)
+	if err != nil {
+		return nil, errors.Wrap(err, "read all keys error")
+	}
+
+	return readAll(keys, b)
+}
+
+func readAll(keys []string, b *RedisDB) (map[string][]byte, error) {
+	result := make(map[string][]byte)
+
+	if len(keys) == 0 {
+		return nil, nil
+	}
+
+	values, err := b.db.MGet(b.ctx, keys...).Result()
+	if err != nil {
+		return nil, errors.Wrap(err, "mget error, read all error")
+	}
+
+	if len(keys) != len(values) {
+		panic("key length does not equal value length")
+	}
+
+	for i, val := range values {
+		byteValue := []byte(fmt.Sprintf("%v", val))
+		key := keys[i]
+		result[key] = byteValue
+	}
+
+	return result, nil
+}
+
+func (b *RedisDB) ReadAllKeys(namespace string) ([]string, error) {
+	return readAllKeys(namespace, b)
+}
+
+func readAllKeys(match string, b *RedisDB) ([]string, error) {
+	var cursor uint64
+
+	allKeys := make([]string, 0)
+
+	for {
+		keys, nextCursor, err := b.db.Scan(b.ctx, cursor, match+"*", RedisScanBatchSize).Result()
+		if err != nil {
+			return nil, errors.Wrap(err, "redis scan error")
+		}
+
+		allKeys = append(allKeys, keys...)
+
+		if nextCursor == 0 {
+			break
+		}
+
+		cursor = nextCursor
+	}
+
+	return allKeys, nil
+}
+
+func (b *RedisDB) Delete(namespace, key string) error {
+	nameSpaceKey := getRedisKey(namespace, key)
+
+	if err := b.mutex.Lock(); err != nil {
+		return errors.Wrap(err, "cannot obtain mutex lock")
+	}
+	defer func() {
+		ok, unlockErr := b.mutex.Unlock()
+		if !ok || unlockErr != nil {
+			logrus.Error(unlockErr)
+		}
+	}()
+
+	return b.db.Del(b.ctx, nameSpaceKey).Err()
+}
+
+func (b *RedisDB) DeleteNamespace(namespace string) error {
+	keys, err := readAllKeys(namespace, b)
+	if err != nil {
+		return errors.Wrap(err, "read all keys error")
+	}
+
+	if err := b.mutex.Lock(); err != nil {
+		return errors.Wrap(err, "cannot obtain mutex lock")
+	}
+	defer func() {
+		ok, unlockErr := b.mutex.Unlock()
+		if !ok || unlockErr != nil {
+			logrus.Error(unlockErr)
+		}
+	}()
+
+	return b.db.Del(b.ctx, keys...).Err()
+}
+
+// TODO: Implement
+func (b *RedisDB) Update(namespace string, key string, values map[string]any) ([]byte, error) {
+	panic("implement me")
+}
+
+// TODO Implement
+func (b *RedisDB) UpdateValueAndOperation(namespace, key string, updater Updater, opNamespace, opKey string, opUpdater ResponseSettingUpdater) (first, op []byte, err error) {
+	panic("implement me")
```

Comment:
This method is actually used for operations, so we shouldn't remove it.

Reply:
Looking at the comment:
// UpdateValueAndOperation updates the value stored in (namespace,key) with the new values specified in the map.
// The updated value is then stored inside the (opNamespace, opKey), and the "done" value is set to true.

It seems the database layer needs to know things about the service

Shouldn't we reimplement this to maybe be like an array of updates to key value pairs and it update both or none at all?

Something like this:
func (b *RedisDB) WriteMany(namespaces []string, key []string, value [][]byte) error

Reply:
Ahh good point, that documentation should be updated. That logic you described is actually encoded in the updater object, so it's specified in the application layer.

Reply:
Note that the Update method is unused, but the UpdateValueAndOperation is.

Reply:
I think this can be removed and we can rework the code to use the newly implemented WriteMany command, what do you think?

func (b *RedisDB) WriteMany(namespaces, keys []string, values [][]byte) error

Reply:
I think that may work, but I'm not sure exactly how. Perhaps you can share what you have in mind? 

I'll note that the existing `UpdateValueAndOperation` takes the value the newly updated value from the first key, and writes that value into the second key. In order to support something like that, we'll need to do a read first and then call WriteMany. That would need to be done within the transaction.

Reply:
I have implemented this function .